### PR TITLE
Better sf support

### PR DESCRIPTION
**Pull request by thomasp85**
This PR adapts `geom_sf` and `coord_sf` to make use of the improvements in performance introduced in sf v0.7-3. The changes are as follows:

- Use `st_normalize` during rescaling to greatly speed up `coord_sf
- Construct a single grob from the geometry and let `sf` decide on how/if to split it up to multiple grobs

This PR introduces a new version dependency on sf as it does not make much sense to support both implementations simultaneously

### CODE REVIEW

**Comment by yutannihilation** [comment]
This fixes #2718 at last. Thanks Thomas!

**Comment by clauswilke** [comment]
This looks good to me in principle, but I don't have time this week to go over it in detail (proofing my book this week, and likely reformatting many of the figures).

I think increasing the required sf version is good anyways, because otherwise axis ticks may be placed in the wrong location due to a pernicious bug that was only recently fixed.

File: R/geom-sf.R [hunk 1]
```diff
@@ -138,33 +130,37 @@ default_aesthetics <- function(type) {
...
-      lty = row$linetype,
-      lineend = lineend,
-      linejoin = linejoin,
-      linemitre = linemitre
-    )
-    sf::st_as_grob(geometry, gp = gp)
-  }
+  geometry <- x$geometry
+  type <- sf_types[sf::st_geometry_type(geometry)]
+  is_point <- type %in% 'point'
```
Comment (yutannihilation) [inline R/geom-sf.R]: Use `"`, not `'`, as the style guide says: https://style.tidyverse.org/syntax.html#quotes

File: R/geom-sf.R [hunk 2]
```diff
@@ -282,3 +278,11 @@ geom_sf_text <- function(mapping = aes(), data = NULL,
     layer_class = LayerSf
   )
 }
+
+sf_types <- c(GEOMETRY = "other", POINT = "point", LINESTRING = "line",
```
Comment (yutannihilation) [inline R/geom-sf.R]: IIUC, the types of simple feature supported by sf can be different in future. Can we just list up geometries of `point` and `line`, and treat others as `others`? (Not sure it works for unknown types, though)

File: R/geom-sf.R [hunk 3]
```diff
@@ -138,33 +130,37 @@ default_aesthetics <- function(type) {
...
-      linejoin = linejoin,
-      linemitre = linemitre
-    )
-    sf::st_as_grob(geometry, gp = gp)
-  }
+  geometry <- x$geometry
+  type <- sf_types[sf::st_geometry_type(geometry)]
+  is_point <- type %in% 'point'
+  type_ind <- match(type, c('point', 'line', 'other'))
+  defaults <- list(
```
Comment (hadley) [inline R/geom-sf.R]: We will need to re-think this at some point in order to expose to the user in the same way as other geoms. But we don't need to do that in this PR.

File: R/geom-sf.R (hunk 1 above)
Comment (thomasp85) [reply R/geom-sf.R]: Oops - nice catch. My life is a constant struggle between my own preference and Hadleys ;-)

File: R/geom-sf.R (hunk 2 above)
Comment (thomasp85) [reply R/geom-sf.R]: This was made in order to get a quick look-up. Most of these are not really supported by sf but just part of the simple feature spec. I'm quite sure sf won't begin to support something other than that in the next 10 years

Comment (yutannihilation) [reply R/geom-sf.R]: OK, it's a good reason. Let's talk about this again in 2029 :)

**Pull request merged**